File: architecture-examples/serverless.js
/**
 * Serverless Architecture - JavaScript
 * 
 * Architecture where code runs in stateless functions triggered by events,
 * managed by cloud providers (AWS Lambda, Azure Functions, etc.)
 */

// Simulated Serverless Function Handler
class ServerlessFunction {
    constructor(name, handler) {
        this.name = name;
        this.handler = handler;
        this.invocations = 0;
        this.coldStarts = 0;
    }
    
    async invoke(event, context) {
        const isColdStart = this.invocations === 0;
        if (isColdStart) {
            this.coldStarts++;
            console.log(`[${this.name}] Cold start - initializing...`);
            await this.initialize();
        }
        
        this.invocations++;
        console.log(`[${this.name}] Invocation #${this.invocations}`);
        
        try {
            const result = await this.handler(event, context);
            return {
                statusCode: 200,
                body: JSON.stringify(result)
            };
        } catch (error) {
            return {
                statusCode: 500,
                body: JSON.stringify({ error: error.message })
            };
        }
    }
    
    async initialize() {
        // Simulate initialization (DB connections, etc.)
        await new Promise(resolve => setTimeout(resolve, 100));
    }
}

// API Gateway Handler
class APIGateway {
    constructor() {
        this.functions = {};
    }
    
    registerFunction(path, method, func) {
        const key = `${method}:${path}`;
        this.functions[key] = func;
    }
    
    async handleRequest(method, path, body) {
        const key = `${method}:${path}`;
        const func = this.functions[key];
        
        if (!func) {
            return {
                statusCode: 404,
                body: JSON.stringify({ error: 'Not found' })
            };
        }
        
        const event = { path, method, body: JSON.parse(body || '{}') };
        const context = { requestId: Date.now().toString() };
        
        return await func.invoke(event, context);
    }
}

// Example Functions
const userHandler = async (event, context) => {
    return {
        message: 'User created',
        userId: Date.now(),
        data: event.body
    };
};

const orderHandler = async (event, context) => {
    return {
        message: 'Order processed',
        orderId: Date.now(),
        data: event.body
    };
};

// Example usage
console.log('=== Serverless Architecture ===\n');

const gateway = new APIGateway();

const createUser = new ServerlessFunction('createUser', userHandler);
const processOrder = new ServerlessFunction('processOrder', orderHandler);

gateway.registerFunction('/users', 'POST', createUser);
gateway.registerFunction('/orders', 'POST', processOrder);

console.log('Invoking functions:');
console.log(await gateway.handleRequest('POST', '/users', JSON.stringify({ name: 'John' })));
console.log(await gateway.handleRequest('POST', '/orders', JSON.stringify({ item: 'Book' })));

module.exports = { ServerlessFunction, APIGateway };

File: architecture-examples/serverless.py
"""
Serverless Architecture - Python

Architecture where code runs in stateless functions triggered by events,
managed by cloud providers (AWS Lambda, Azure Functions, etc.)
"""

import json
import asyncio
from typing import Dict, Callable, Any


# Simulated Serverless Function Handler
class ServerlessFunction:
    def __init__(self, name: str, handler: Callable):
        self.name = name
        self.handler = handler
        self.invocations = 0
        self.cold_starts = 0
    
    async def invoke(self, event: Dict, context: Dict) -> Dict:
        is_cold_start = self.invocations == 0
        if is_cold_start:
            self.cold_starts += 1
            print(f"[{self.name}] Cold start - initializing...")
            await self.initialize()
        
        self.invocations += 1
        print(f"[{self.name}] Invocation #{self.invocations}")
        
        try:
            result = await self.handler(event, context)
            return {
                'statusCode': 200,
                'body': json.dumps(result)
            }
        except Exception as error:
            return {
                'statusCode': 500,
                'body': json.dumps({'error': str(error)})
            }
    
    async def initialize(self):
        # Simulate initialization (DB connections, etc.)
        await asyncio.sleep(0.1)


# API Gateway Handler
class APIGateway:
    def __init__(self):
        self.functions: Dict[str, ServerlessFunction] = {}
    
    def register_function(self, path: str, method: str, func: ServerlessFunction):
        key = f"{method}:{path}"
        self.functions[key] = func
    
    async def handle_request(self, method: str, path: str, body: str = None) -> Dict:
        key = f"{method}:{path}"
        func = self.functions.get(key)
        
        if not func:
            return {
                'statusCode': 404,
                'body': json.dumps({'error': 'Not found'})
            }
        
        event = {'path': path, 'method': method, 'body': json.loads(body or '{}')}
        context = {'requestId': str(int(__import__('time').time()))}
        
        return await func.invoke(event, context)


# Example Functions
async def user_handler(event: Dict, context: Dict) -> Dict:
    return {
        'message': 'User created',
        'userId': int(__import__('time').time()),
        'data': event['body']
    }


async def order_handler(event: Dict, context: Dict) -> Dict:
    return {
        'message': 'Order processed',
        'orderId': int(__import__('time').time()),
        'data': event['body']
    }


# Example usage
async def main():
    print("=== Serverless Architecture ===\n")
    
    gateway = APIGateway()
    
    create_user = ServerlessFunction('createUser', user_handler)
    process_order = ServerlessFunction('processOrder', order_handler)
    
    gateway.register_function('/users', 'POST', create_user)
    gateway.register_function('/orders', 'POST', process_order)
    
    print("Invoking functions:")
    print(await gateway.handle_request('POST', '/users', json.dumps({'name': 'John'})))
    print(await gateway.handle_request('POST', '/orders', json.dumps({'item': 'Book'})))


if __name__ == "__main__":
    asyncio.run(main())

